candidate_host.go

```diff
@@ -12,26 +12,42 @@ type CandidateHost struct {
 	network string
 }
 
+// CandidateHostConfig is the config required to create a new CandidateHost
+type CandidateHostConfig struct {
+	CandidateID *string
+	Network     string
+	Address     string
+	Port        int
+	Component   uint16
+}
+
 // NewCandidateHost creates a new host candidate
-func NewCandidateHost(network string, address string, port int, component uint16) (*CandidateHost, error) {
-	candidateID, err := generateCandidateID()
-	if err != nil {
-		return nil, err
+func NewCandidateHost(config *CandidateHostConfig) (*CandidateHost, error) {
+	var candidateID string
+	var err error
+
+	if config.CandidateID != nil {
```

Comment:
Instead of a `*` I would use an empty string to force generation (is there ever a case where we want an empty string ID?)